README.md
An unofficial API for [royalroadl.com](https://royalroadl.com), written in TypeScript.

```
npm i -s @l1lly/royalroadl-api
```

This is an attempt to write a predictable and consistent wrapper around the  mess that is RRL. Since no official public API is exposed, this module scrapes all data straight from the HTML, which makes it very prone to spontaneous and horrible death.

Barebones documentation generated from [TypeDoc](http://typedoc.org/) can be found on [fsoc.gitlab.com/royalroadl-api](https://fsoc.gitlab.io/royalroadl-api/classes/royalroadapi.html). You can also build these docs yourself by running `npm run docs` in the root of the project.

A more elaborate description of this package and its functionality (both internal and exposed) can be found further below in the [about](#about) section.

## Example usage

For more examples check out the `/examples` directory.

```javascript
const { RoyalRoadAPI } = require('@l1lly/royalroadl-api');

const api = new RoyalRoadAPI();

(async () => {

const { data } = await api.fictions.getPopular();
const titles = data.slice(10).map((fic) => fic.title);

console.log(`The top 10 popular fictions are: ${titles.join(', ')}`);

})().catch(console.error);
```

## About

The module itself exports only a `RoyalRoadAPI` class which, by itself, has no methods. All functionality is delegated to service classes which are properties of the `RoyalRoadAPI` instance. This allows for a very concise seperation of concerns, and a modular and atomic approach to both the development, and usage of this module.

### Responses

All responses and errors are either an instance of a `RoyalResponse` or an `RoyalError`, which extends `RoyalResponse`. This is done to easily allow for meta information to be tacked onto responses, and to have a consistent interface between user and module. Note that the `RoyalError` acts similarly to the NodeJS `Error` object, in that it captures and returns a short stack trace.

### Internal requester

All service classes use the same instance of the `Requester`, the class responsible for making HTTP requests and returning their responses. By default, it will throw a `RoyalError` if it encounters a status code other than 200 (this can be disabled with the `ignoreStatus` option). Also note that, by default, all requests will be sent over HTTPS - this can be controlled with the first argument of the `RoyalRoadAPI` constructor.

Since RRL likes to return 200 even when the actual response should be a 404 or 304, the `Requester` will parse the HTML it has gotten (if it got any), and try to read an error from it. If it finds signs that the request has failed, it will throw - this can be disabled with the `ignoreParser` option.

The `Requester`s main goal is to keep track of cookies and to automatically fetch a `__ResponseVerificationToken` often needed for POST requests as a part of anti CSRF measures. This fetching of tokens is disabled by default and can be enabled with the `fetchToken` option.

### Structure

All services are structured in a very similar way: with a `<Type>Service` exposing all relevant methods, and a `<Type>Parser`, which usually exposes a number of static methods used to parse HTML responses.

A quick wrap-up of all the existing services is: 
- `ChapterService`, contains methods for fetching and publishing chapters and chapter comments.
- `FictionService`, fetching fiction data and reviews.
- `FictionsService`, methods for fetching all types of fiction lists RRL offers, with their respective levels of per fiction detail.
- `ProfileService`, handling profiles, returns parsed user profiles.
- `UserService`, actions related to the logged-in user like logon, getting the users' fictions, bookmarks, or notifications.

### Parsing

This uses [cheerio](https://github.com/cheeriojs/cheerio) to parse HTML, which is a very forgiving parser. This means that even if RRL were to make minor changes to their page layouts, large parts of the API (even those parts responsible for changed areas) would still remain functional.

Therefore, expect properties to be empty or `null`, and know that an error will not be thrown just because some values could not be parsed.

CHANGELOG.md
### 0.8.0

- implement chapter publishing
- implement request verification token fetching for POST requests

### 0.7.0

- expose only relevant code to NPM
- implement Requester class, update code for new interface

### 0.6.1

- move types to the files they are primarily being used in

### 0.6.0

- implement UserService

src/royalroad.ts
import * as logger from 'debug';
import * as cheerio from 'cheerio';
import * as request from 'request';

import { RoyalError } from './responses';
import { UserService } from './services/user';
import { ProfileService } from './services/profile';
import { ChapterService } from './services/chapter';
import { FictionService } from './services/fiction';
import { FictionsService } from './services/fictions';
import { getBaseAddress, getUserAgent } from './constants';

interface RequestOptions {
  fetchToken?: boolean;
  ignoreStatus?: boolean;
  ignoreParser?: boolean;
}

/**
 * Class passed to all Services for consistent cookies accross requests.
 */
export class Requester {
  private static readonly headers = {
    'User-Agent': getUserAgent(),
  };

  public cookies: any; // TODO: Dangerous!
  public insecure: boolean;

  private readonly url: string;
  private readonly debug: logger.IDebugger;

  constructor(insecure = false) {
    this.insecure = insecure;
    this.cookies = request.jar();
    this.debug = logger('requester');
    this.url = getBaseAddress(insecure);
  }

  get isAuthenticated() {
    const cookies = this.cookies.getCookies(
      getBaseAddress(this.insecure),
    );

    for (const cookie of cookies) {
      // TODO: Dangerous.
      const c = cookie as any as { key: string, value: string };

      if (c.key === 'mybbuser' && c.value) {
        return true;
      }
    }

    return false;
  }

  /**
   * GET request to the given path, under the base address.
   *
   * @param path
   */
  public async get(path: string) {
    const uri = this.url + path;

    const body = await this.request({
      uri,
      jar: this.cookies,
      headers: Requester.headers,
    });

    return body;
  }

  /**
   * POST request to the given path, under the base address.
   *
   * @param path
   */
  public async post(path: string, data: any, options: RequestOptions = {}) {
    const uri = this.url + path;

    data['__RequestVerificationToken'] = options.fetchToken ? (
      await this.fetchToken(path)
    ) : undefined;

    this.debug(data);

    const body = await this.request({
      uri,
      form: data,
      method: 'POST',
      jar: this.cookies,
    }, options);

    return body;
  }

  private request(
    req: request.UriOptions & request.CoreOptions, options: RequestOptions = {},
  ): Promise<string> { return new Promise((resolve, reject) => {
    this.debug('%o: %o', req.method || 'GET', req.uri);

    request(req, (err, res, body) => {
      if (err || (res.statusCode !== 200 && !options.ignoreStatus)) {
        this.logCookies();

        return reject(new RoyalError(
          err ? err.message || err : res.statusMessage || 'Requester error',
        ));
      }

      this.debug('%o < %o (%o)',
        res.method || 'GET', res.statusCode, res.statusMessage);

      if (!options.ignoreParser) {
        const genericError = this.catchGenericError(body);
        if (genericError !== null) {
          reject(new RoyalError(genericError));
        }
      }

      resolve(body);
    });
  }); }

  /**
   * Log all cookies in the jar, shorten values which are longer than
   * 20 characters.
   *
   * @param insecure
   */
  private logCookies(insecure?: boolean) {
    const addr = getBaseAddress(insecure);
    const cookies = this.cookies.getCookies(addr)
      .filter((c: any) => c)
      .map((c: any) => `cookie ${c.key} = ${c.value.length >= 18
        ? c.value.slice(0, 17) + '...' : c.value}`);

    this.debug(cookies);
  }

  /**
   * Fetch a request verification token from a hidden input on the
   * target path. Throws if it couldn't find one.
   *
   * @param path
   */
  private async fetchToken(path: string) {
    const body = await this.get(path);
    const $ = cheerio.load(body);

    const genericError = this.catchGenericError(body);
    if (genericError !== null) {
      throw new RoyalError(genericError);
    }

    const token: string = $('input[name="__RequestVerificationToken"]')
      .prop('value');

    this.debug('got token %o', token);

    if (token && token.length !== 0) {
      return token;
    } else { throw new RoyalError('Token not found.'); }
  }

  /**
   * Catch generic errors like 404 and 403, since RRL always returns with status
   * code 200.
   *
   * @param html
   */
  private catchGenericError(html: string) {
    const $ = cheerio.load(html);
    const message = $('div.page-404').find('h3').text().trim();

    return (message && message.length !== 0) ? message : null;
  }
}

/**
 * Container class, creating instances of the seperate Service classses.
 */
export class RoyalRoadAPI {
  public readonly user: UserService;
  public readonly profile: ProfileService;
  public readonly chapter: ChapterService;
  public readonly fiction: FictionService;
  public readonly fictions: FictionsService;

  private readonly req: Requester;

  constructor(insecure: boolean = false) {
    this.req = new Requester(insecure);

    this.user = new UserService(this.req);
    this.profile = new ProfileService(this.req);
    this.chapter = new ChapterService(this.req);
    this.fiction = new FictionService(this.req);
    this.fictions = new FictionsService(this.req);
  }
}
